### PR TITLE
THREESCALE-5894  Hpa feature branch merge

### PR DESCRIPTION
**Pull request by austincunningham**
## What
[THREESCALE-5894](https://issues.redhat.com/browse/THREESCALE-5894) enable hpa for backend listener, worker and apicast production

### Positive case 
Async mode enabled 
We will need non logical backend redis to enable async mode 
- [ ] you can Run this [script](https://github.com/3scale-labs/3scale-perf-setup/blob/main/setup_db.sh) with `./setup_db.sh` to setup the db and redis and create the db secrets in the 3scale ns

- [ ] from the 3scale-operator repo run `make install`
- [ ] `make run`
- [ ] Create a s3 secret
```bash
kubectl apply -f - <<EOF
---
kind: Secret
apiVersion: v1
metadata:
  name: s3-credentials
data:
  AWS_ACCESS_KEY_ID: UkVQTEFDRV9NRQ==
  AWS_BUCKET: UkVQTEFDRV9NRQ==
  AWS_REGION: UkVQTEFDRV9NRQ==
  AWS_SECRET_ACCESS_KEY: UkVQTEFDRV9NRQ==
type: Opaque
EOF
```
- [ ] get the default application router
```bash
DOMAIN=$(oc get routes console -n openshift-console -o json | jq -r '.status.ingress[0].routerCanonicalHostname' | sed 's/router-default.//')
```
- [ ] Create an apimanager CR using that secret e.g.
```bash
kubectl apply -f - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager-sample
spec:
  system:
    database:
      postgresql: {}
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: s3-credentials
  wildcardDomain: $DOMAIN
EOF
```
- [ ] once your installation is finished set the following fields in the apmanager cr to true
- `spec.apicast.productionSpec.hpa: true`
- `spec.backend.workerSpec.hpa: true`
- `spec.backend.listenerSpec.hap: true`

- [ ] confirm that 3 hpa objects are created with the default spec
```bash
oc get hpa
NAME                 REFERENCE                             TARGETS            MINPODS   MAXPODS   REPLICAS   AGE
apicast-production   DeploymentConfig/apicast-production   54%/85%, 0%/85%    1         5         1          69m
backend-listener     DeploymentConfig/backend-listener     13%/85%, 0%/85%    1         5         1          69m
backend-worker       DeploymentConfig/backend-worker       108%/85%, 0%/85%   1         5         5          69m
```

- Confirm that a warning condition is visible on the apimanager CR indicating that requests, limits, and replicas values are ignored because HPA took over.
```yaml
status:
  conditions:
    - lastTransitionTime: '2024-01-18T10:00:33Z'
      status: 'True'
      type: Available
    - lastTransitionTime: '2024-01-17T12:22:58Z'
      message: >-
        HorizontalPodAutoscaling (Hpa) enabled overrides values applied to
        request, limits and replicas
      reason: HPA
      status: 'True'
      type: Warning
```
- Confirm that requests and limits are set to the same values  in the apicast-production, backend-listener and backend-worker deployments
#### Confirm Async is enabled 
- [ ] Check the backend worker pod for the `CONFIG_REDIS_ASYNC` envar
```bash
$ oc get po | grep backend-worker
backend-worker-1-deploy       0/1     Completed   0             12m
backend-worker-1-jn29j        1/1     Running     0             12m
$ oc exec backend-worker-1-jn29j -- env | grep CONFIG_REDIS_ASYNC
Defaulted container "backend-worker" out of: backend-worker, backend-redis-svc (init)
CONFIG_REDIS_ASYNC=1
```
- [ ] Check the backend listener pod for the `CONFIG_REDIS_ASYNC` envar and `LISTENER_WORKERS`
```bash
$ oc get po | grep backend-listener
backend-listener-1-deploy       0/1     Completed   0             12m
backend-listener-1-jn29j        1/1     Running     0             12m
$ oc exec backend-listener-1-jn29j -- env | grep CONFIG_REDIS_ASYNC
Defaulted container "backend-listener" out of: backend-listener, backend-redis-svc (init)
CONFIG_REDIS_ASYNC=1
$ oc exec backend-listener-1-jn29j -- env | grep LISTENER_WORKERS
Defaulted container "backend-listener" out of: backend-listener, backend-redis-svc (init)
LISTENER_WORKERS=1
```
- [ ] Check the backend listener for the `falcon` Arg
```bash
$ oc get deploy backend-listener -oyaml | yq '.spec.template.spec.containers[0].args'
- bin/3scale_backend
- -s
- falcon
- start
- -e
- production
- -p
- "3000"
- -x
- /dev/stdout
```

- [ ] remove the envar from the backend-worker  and backend-listener deployment config `spec.containers.env` and see if its reconciled back e.g. remove this
```yaml
- name: CONFIG_REDIS_ASYNC
  value: '1'
```
- [ ] set the value in env var CONFIG_REDIS_ASYNC to '0' and it should revert to '1'

### Deletion

- [ ] delete or set to false the following
- `spec.apicast.productionSpec.hpa: true`
- `spec.backend.workerSpec.hpa: true`
- `spec.backend.listenerSpec.hap: true`
- [ ] Confirm that the hpa objects are deleted from the cluster
```bash
oc get hpa
```

### Negative case
Async mode disabled for backend worker and listener

- [ ] Create a project `oc new-project 3scale-test2`
- [ ] `make install`
- [ ] `make run`
- [ ] Create a s3 secret
```bash
kubectl apply -f - <<EOF
---
kind: Secret
apiVersion: v1
metadata:
  name: s3-credentials
data:
  AWS_ACCESS_KEY_ID: UkVQTEFDRV9NRQ==
  AWS_BUCKET: UkVQTEFDRV9NRQ==
  AWS_REGION: UkVQTEFDRV9NRQ==
  AWS_SECRET_ACCESS_KEY: UkVQTEFDRV9NRQ==
type: Opaque
EOF
```
- [ ] get the default application router
```bash
DOMAIN=$(oc get routes console -n openshift-console -o json | jq -r '.status.ingress[0].routerCanonicalHostname' | sed 's/router-default.//')
```
- [ ] Create an apimanager CR using that secret e.g.
```bash
kubectl apply -f - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager-sample
spec:
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: s3-credentials
  wildcardDomain: $DOMAIN
EOF
```
- [ ] once your installation is finished set the following fields in the apmanager cr to true
- `spec.apicast.productionSpec.hpa: true`
- `spec.backend.workerSpec.hpa: true`
- `spec.backend.listenerSpec.hpa: true`

- [ ] confirm that only the hpa objects for apicast-production is created with the default spec
```bash
oc get hpa
NAME                 REFERENCE                             TARGETS            MINPODS   MAXPODS   REPLICAS   AGE
apicast-production   DeploymentConfig/apicast-production   54%/85%, 0%/85%    1         5         1          69m
```
- [ ] Confirm that a warning condition is visible on the apimanager CR indicating that requests, limits, and replicas values are ignored because HPA took over.
```yaml
status:
  conditions:
    - lastTransitionTime: '2024-01-17T12:29:42Z'
      status: 'False'
      type: Available
    - lastTransitionTime: '2024-01-17T12:36:09Z'
      message: >-
        HorizontalPodAutoscaling (Hpa) Logical Redis databases detected for
        backend, logical Redis database not compatible with async mode, Hpa
        requires async mode for backend, Hpa not enabled for backend
      reason: HPA
      status: 'True'
      type: Configuration Warning
    - lastTransitionTime: '2024-01-17T14:02:26Z'
      message: >-
        HorizontalPodAutoscaling (Hpa) enabled overrides values applied to
        request, limits and replicas
      reason: HPA
      status: 'True'
      type: Warning
```

### CODE REVIEW

**Review by carlkyrillos — COMMENTED** [review]
I added a few nitpicky comments about changing the import order to stick with the project convention. I didn't comment on every file but most of the `.go` files could have their import order tidied up.

The actual code looks good to me. I verified both the positive and negative cases and everything worked without issue.

File: pkg/3scale/amp/component/backend_monitoring.go [hunk 1]
```diff
@@ -2,7 +2,6 @@ package component
 
 import (
 	"fmt"
-
```
[the remark this answers is not in the document]
Comment (austincunningham) [reply pkg/3scale/amp/component/backend_monitoring.go:5]: Sure generally leave imports to the ide to format.  Will change and hopefully the ide doesn't revert the change.

File: pkg/3scale/amp/component/hpa.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func DefaultHpa(name string, namespace string) *hpa.HorizontalPodAutoscaler {
+	minPods := helper.Int32Ptr(1)
+	maxPods := int32(5)
+	cpuPercent := helper.Int32Ptr(85)
+	memoryPercent := helper.Int32Ptr(85)
+
+	// needs to change from dc to deployment
```
[the remark this answers is not in the document]
Comment (austincunningham) [reply pkg/3scale/amp/component/hpa.go:15]: yes

File: pkg/3scale/amp/operator/backend_reconciler.go [hunk 3]
```diff
@@ -91,6 +101,10 @@ func (r *BackendReconciler) Reconcile() (reconcile.Result, error) {
 
 	// Worker Deployment
 	workerDeploymentMutator := reconcilers.GenericBackendDeploymentMutators()
+	//workerConfigMutator := reconcilers.GenericBackendMutators()
```
[the remark this answers is not in the document]
Comment (austincunningham) [reply pkg/3scale/amp/operator/backend_reconciler.go:104]: yes

**Comment by carlkyrillos** [comment]
Reviewed latest changes and re-verified the positive and negative cases
/lgtm

**Pull request merged**